<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Promise Async-await</title>
</head>
<body>
    <h1>Q1. Write one example explaining how you can write a callback function.</h1>

    <pre style="color:blue"> function calculate(n1, n2, callback){
        const result = n1 + n2;
        callback(result);
      }
      function printResult(output){
        console.log(`The result is: ${output}`);
      }
      calculate(10, 25, printResult)</pre>
   
    <p>There are two functions named calculate() and printResult(). calculate() has three parameters and printResult() has one parameter. At last 1st function has been called
        with 3 arguments within that printResult function has been taken as argument. 
    </p>   
    <hr>

    <h1>Q2. Write a callback function to print numbers from 1 to 7, in which 1 should be printed after 1 sec, 2 should be printed after 2 sec, 3 should be printed after 3 sec, and so on. 

Explain callback hell.
Numbers
1
2
3
4
5
6
7
</h1>
    <pre style="color:blue">function printNum(time) {
        let count = 0;
        for(let i=1;i< 8;i++){
        setTimeout(() => {        
            console.log(i)
        },(count)* time );
        count+=i
    } 
    }
    printNum(1000)</pre>

    <p> Callback hell- Callback hell is a nested function, which has functions within the functions and when we execute that function due to some error in one function 
        the desired output may not be delivered. Its like a hell to debug the code going in a function within a function.
    </p> 
     <hr>

    <h1>Q3.Write the promise function to print numbers from 1 to 7, in which 1 should be printed after 1 sec, 2 should be printed after 2 sec, 3 should be printed after 3 sec, and so on. 

Numbers
1
2
3
4
5
6
7
</h1>

    <pre style="color:blue">
    function resolvePromise(alpha, timeout){
        return new Promise((resolve, reject) => {
        setTimeout(() => {
        console.log(alpha)
        resolve("Promise is Resolved Successfully...");
        }, timeout);
        })
        }
        async function vowelAlpha(){
        await resolvePromise('A', 2000) 
        await resolvePromise("E", 3000)
        await resolvePromise("I", 4000)
        await resolvePromise("O", 2000)
        await resolvePromise("U", 7000)
        }
        vowelAlpha();</pre>
    <!-- <pre style="color:blue">function count(nums,time){
        return new Promise((res,rej) =>{
            if(nums > 0){
                setTimeout(() => {
                    res(nums++);
                }, time);
            }
            else{
                rej("Number Should be greater than 0")
            }
        })
    }
    
    count(0,1000)
    .then((data) =>console.log(data))
    .catch((err) =>console.log(err))
    </pre> -->
    <p></p> 
     <hr>
     <h1>Q4. Create a promise function accepting an argument, if yes is passed to the function then it should go to resolved state and print Promise Resolved, and if nothing is passed 
        then it should go to reject the state and catch the error and print Promise Rejected 
        </h1>
        <pre style="color:blue">
            function printNumber(N) {
                let promise = new Promise ((resolve, reject) => {
                   
                    if (N){
                        resolve(`Promise resolved`)
                    } else{
                        reject(`Promise rejected`)
                    }
                })
                
                
                promise
                .then((yes) => console.log(yes))
                .catch((err) => console.log(err))
                }
                printNumber("yes")
        </pre>
        <hr>

    <h1>Q5. Create examples to explain callback function</h1>
    
    <pre style="color:blue">function abc() {
        let a = 1
        function xyz() {
            console.log(a);
        }
        xyz()
    }
    abc()</pre>

    <pre style="color:blue">setTimeout(() => {
        setTimeout(() => {
            setTimeout(() => {
                console.log("no");
            }, 3000);
        }, 2000);
    }, 1000);</pre>

    <p> Callback function are the functions where one function is dependent on another function, when one of the function is called its argument is passed as parametr to
        amother function. In another words when one function is called the another function is callbacked to execute. </p> 
     <hr>

    <h1>Q6. Create examples to explain callback hell function</h1>
    <pre style="color:blue">setTimeout(() => {
        console.log("step 1");
        setTimeout(() => {
            console.log("step 2");
            setTimeout(() => {
                console.log("step 3");
                setTimeout(() => {
                    console.log("step 4");
                       setTimeout(() => {
                          console.log("step 5");
                        setTimeout(() => {
                            console.log("step 7");
                        },8000);
                        setTimeout(() => {
                            console.log("step 6");
                        },6000);
                    },7000);
                },4000);
            }, 3000);
        }, 2000);
    }, 1000);
    </pre>
    <p> Callback hell- Callback hell is a nested function, which has functions within the functions and when we execute that function due to some error in one function 
        the desired output may not be delivered. Its like a hell to debug the code going in a function within a function.
    </p> 

     <hr>


    <h1>Q7. Create examples to explain promises function</h1>
    <pre style="color:blue">let a = new Promise ((resolve, reject) =>{
        let  number = N;
        let count = 0;
        for(i=1;i< =N;i++){
            if(N%i===0){
                count++;
            }
        }
        if(count===2){
            resolve(`${N} is Prime Number`)
        }else{
            reject(`${N} is Not a Prime Number`)
        }
    }) 
    
    a
    .then((Prime)=> console.log(Prime))
    .catch((error)=>console.log(error))</pre>

    <p> A promise has two parameters resolve and reject which are also functions. Resolve is like the true statement and reject is like a false statement. when the condition
        is true then it executes resolve statement taking ".then" function . if statement is false it gets rejected and executes ".catch" statement</p> 
     <hr>

    <h1>Q8. Create examples to explain async await function</h1>
    <pre style="color:blue">function resolvePromise(alpha, timeout){
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                console.log(alpha)
            resolve("Promise is Resolved Successfully...");
            }, timeout);
        })
      }
     async function vowelAlpha(){
        await resolvePromise('A', 2000)
        await resolvePromise("E", 3000)
        await resolvePromise("I", 4000)
        await resolvePromise("O", 2000)
        await resolvePromise("U", 7000)
    }
    vowelAlpha();
    </pre>
    <p>Async await function uses asyncand awaiit keywords. Normally we have created a promise function then we have use async keyword for the function vowelAlpha() within it we 
        used await keyword which says to wait till that line executes after that the next line can execute. 
    </p> 
    <hr>

    <h1>Q9. Create examples to explain promise.all function</h1>
    <pre style="color:blue">
        const mapData = () =>
          new Promise((resolve,reject) => {
          console.log("fetch map data")
          setTimeout(() => resolve(), 1000)
        });
        const weatherData= () =>
          new Promise((resolve,reject) => {
          console.log("fetch wether data")
          setTimeout(() => resolve(),1000)
        });
  
        const collectAll = () => console.log("after both Mapdata and weatherdata are collected")
  
        Promise.all([mapData(), weatherData()]).then(() => {
           collectAll()
        });</pre>
    <p>Promise all function makes all the function to execute irrespective of conections or dependability on each other .</p> 
    <script src="./promise.js"></script>
</body>
</html>